cds/utils.bal

```diff
@@ -0,0 +1,192 @@
+            }
+        }
+    } else {
+        foreach var key in prefetchData.keys() {
+            r4:FHIRValidationError? validated = validator:validate(prefetchData.get(key));
+            if validated is r4:FHIRValidationError {
+                return createCdsError(string `FHIR data provided for the prefetch: ${key}`, 400, cause = validated);
+            }
+        }
+    }
+
+    return request;
+}
+
+isolated function fetchFhirResource(string prefetchTemplateKey, string prefetchTemplate, CdsRequest cdsRequest) returns CdsError|CdsRequest {
+    string template = prefetchTemplate.clone();
+    string split = regexp:split(re `\.`, regexp:split(re `\}\}`, regexp:split(re `\{\{`, prefetchTemplate)[1])[0])[1];
+    string fhirServer = <string>cdsRequest.fhirServer;
+    anydata contextValue = cdsRequest.context.get(split);
+    if contextValue is string {
+        template = regexp:replace(re `\{\{context.${split}\}\}`, template, contextValue);
+    }
+
+    if (fhirServer.endsWith("/")) {
+        fhirServer = fhirServer.substring(0, fhirServer.length() - 1);
+    }
+
+    if (!(template.startsWith("/"))) {
+        template = string `/${template}`;
+    }
+
+    http:Client|http:ClientError fhirClient = new (fhirServer);
+    if (fhirClient is http:Client) {
+        map<string|string[]> headers = {
+            "Authorization": string `Bearer ${(<FhirAuthorization>cdsRequest.fhirAuthorization).access_token}`
+        };
+        http:Response|http:ClientError response = fhirClient->get(template, headers);
+        if (response is http:Response) {
+            json|http:ClientError jsonPayload = response.getJsonPayload();
+            if (jsonPayload is http:ClientError) {
+                return createCdsError(string `FHIR data retrieved for : ${fhirServer}${template} is not JSON`, 412, cause = jsonPayload);
+            } else {
+                r4:DomainResource|error parsedResource = parser:parse(jsonPayload).ensureType(r4:DomainResource);
+                if parsedResource is r4:DomainResource {
+                    if (parsedResource is r4:OperationOutcome) {
+                        return createCdsError(string `FHIR data retrieved for : ${fhirServer}${template} is not valid`, 412);
+                    } else {
+                        r4:FHIRValidationError? validatedData = validator:validate(jsonPayload);
+                        if validatedData is r4:FHIRValidationError {
+                            return createCdsError(string `FHIR data retrieved for : ${fhirServer}${template} is not valid`, 412, cause = validatedData);
+                        } else {
+                            cdsRequest.prefetch[prefetchTemplateKey] = parsedResource;
+                            return cdsRequest;
+                        }
+                    }
+                } else {
+                    return createCdsError(string `Data retrieved for : ${fhirServer}${template} is not FHIR data`, 412, cause = parsedResource);
+                }
```

Comment:
Apply early returns to avoid nested if-else much as possible.
https://learn-ballerina.github.io/best_practices/early_returns.html

Reply:
Done